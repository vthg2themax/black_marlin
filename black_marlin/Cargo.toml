[package]
name = "black_marlin"
version = "0.3.4"
authors = ["Ryohei Machida <orcinus4627@gmail.com>"]
description = "Simple, small, and extremely fast template engine for Rust. Forked from SailFish, now it's a black marlin."
homepage = "https://github.com/vthg2themax/black_marlin"
repository = "https://github.com/vthg2themax/black_marlin"
readme = "../README.md"
keywords = ["markup", "template", "html", "classic", "class-asp","easy html generation","simplet"]
categories = ["template-engine"]
license = "MIT"
workspace = ".."
edition = "2018"

[features]
default = ["derive", "perf-inline"]
# automatically import derive macro
derive = ["black_marlin-macros"]
# enable json filter
json = ["serde", "serde_json"]
# add more #[inline] attribute
perf-inline = []

[dependencies]
itoap = "1.0.1"
ryu = "1.0.9"
serde = { version = "1.0.133", optional = true }
serde_json = { version = "1.0.74", optional = true }

[dependencies.black_marlin-macros]
path = "../black_marlin-macros"
version = "0.3.4"
optional = true

[build-dependencies]
version_check = "0.9.4"

[package.metadata.docs.rs]
all-features = true
rustdoc-args = ["--cfg", "docsrs"]