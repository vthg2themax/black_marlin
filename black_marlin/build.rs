fn main() {
    if version_check::is_feature_flaggable() == Some(true) {
        println!("cargo:rustc-cfg=black_marlin_nightly");
    }
}
