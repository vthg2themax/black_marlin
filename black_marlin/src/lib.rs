//! black_marlin is a simple, small, and extremely fast template engine for Rust.
//! Before reading this reference,
//! I recommend reading [User guide](https://sailfish.netlify.app/en/).
//!
//! This crate contains utilities for rendering black_marlin templates.
//! If you want to use black_marlin templates, import `black_marlin-macros` crate and use
//! derive macro `#[derive(TemplateOnce)]` or `#[derive(Template)]`.
//!
//! In most cases you don't need to care about the `runtime` module in this crate, but
//! if you want to render custom data inside templates, you must implement
//! `runtime::Render` trait for that type.
//!
//! ```ignore
//! use black_marlin::TemplateOnce;
//!
//! #[derive(TemplateOnce)]
//! #[template(path = "hello.stpl")]
//! struct HelloTemplate {
//!     messages: Vec<String>
//! }
//!
//! fn main() {
//!     let ctx = HelloTemplate {
//!         messages: vec!["foo".to_string(), "bar".to_string()]
//!     };
//!
//!     println!("{}", ctx.render_once().unwrap());
//! }
//! ```

#![doc(
    html_logo_url = "https://raw.githubusercontent.com/Kogia-sima/sailfish/master/resources/icon.png"
)]
#![cfg_attr(black_marlin_nightly, feature(core_intrinsics))]
#![cfg_attr(docsrs, feature(doc_cfg))]
#![allow(clippy::redundant_closure)]
#![deny(missing_docs)]

pub mod runtime;

use runtime::Buffer;
pub use runtime::{RenderError, RenderResult};
#[cfg(feature = "derive")]
#[cfg_attr(docsrs, doc(cfg(feature = "derive")))]
pub use black_marlin_macros::TemplateOnce;

/// Template that can be rendered with consuming itself.
pub trait TemplateOnce: Sized + private::Sealed {
    /// Render the template and return the rendering result as `RenderResult`
    ///
    /// This method never returns `Err`, unless you explicitly return RenderError
    /// inside templates
    ///
    /// When you use `render_once` method, total rendered size will be cached, and at
    /// the next time, buffer will be pre-allocated based on the cached length.
    ///
    /// If you don't want this behaviour, you can use `render_once_to` method instead.
    fn render_once(self) -> runtime::RenderResult;

    /// Render the template and append the result to `buf`.
    ///
    /// This method never returns `Err`, unless you explicitly return RenderError
    /// inside templates
    ///
    /// ```
    /// use black_marlin::TemplateOnce;
    /// use black_marlin::runtime::Buffer;
    ///
    /// # pub struct HelloTemplate {
    /// #   messages: Vec<String>,
    /// # }
    /// #
    /// # impl TemplateOnce for HelloTemplate {
    /// #     fn render_once(self) -> Result<String, black_marlin::RenderError> {
    /// #         Ok(String::new())
    /// #     }
    /// #
    /// #     fn render_once_to(self, buf: &mut Buffer)
    /// #             -> Result<(), black_marlin::RenderError> {
    /// #         Ok(())
    /// #     }
    /// # }
    /// #
    /// # impl black_marlin::private::Sealed for HelloTemplate {}
    /// #
    /// let tpl = HelloTemplate {
    ///     messages: vec!["foo".to_string()]
    /// };
    ///
    /// // custom pre-allocation
    /// let mut buffer = Buffer::with_capacity(100);
    /// tpl.render_once_to(&mut buffer).unwrap();
    /// ```
    fn render_once_to(self, buf: &mut Buffer) -> Result<(), RenderError>;
}

/// Work in Progress
pub trait Template: private::Sealed {
    /// Work in progress
    fn render(&self) -> runtime::RenderResult;
}

#[doc(hidden)]
pub mod private {
    pub trait Sealed {}
}
