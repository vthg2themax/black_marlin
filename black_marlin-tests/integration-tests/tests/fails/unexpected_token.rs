use black_marlin::TemplateOnce;
use black_marlin_macros::TemplateOnce;

struct Content<'a> {
    id: u32,
    address: &'a str,
    phone_number: &'a str,
}

#[derive(TemplateOnce)]
#[template(path = "unexpected_token.stpl")]
#[template(escape = false)]
struct UnexpectedToken<'a> {
    name: &'a str,
    content: Content<'a>
}

fn main() {
    println!(
        "{}",
        UnclosedToken {
            name: "Taro",
            content: Content {
                id: 1,
                address: "oooo-xxxx",
                phone_number: "000-000-0000"
            }
        }
        .render_once()
        .unwrap()
    )
}
