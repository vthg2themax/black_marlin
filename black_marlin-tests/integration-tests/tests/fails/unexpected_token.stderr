error: Failed to compile template.
caused by: Rust Syntax Error (expected expression)

file: unexpected_token.stpl
position: line 3, column 17

  |
3 |   "content": <% =content %>
  |                 ^

  --> $DIR/unexpected_token.rs:10:10
   |
10 | #[derive(TemplateOnce)]
   |          ^^^^^^^^^^^^
   |
   = note: this error originates in a derive macro (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0422]: cannot find struct, variant or union type `UnclosedToken` in this scope
  --> $DIR/unexpected_token.rs:21:9
   |
21 |         UnclosedToken {
   |         ^^^^^^^^^^^^^ not found in this scope

warning: unused import: `sailfish::TemplateOnce`
 --> $DIR/unexpected_token.rs:1:5
  |
1 | use sailfish::TemplateOnce;
  |     ^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: `#[warn(unused_imports)]` on by default
